src/main/java/edu/texas/social_computing/hospitals/RHRP.java

```diff
@@ -0,0 +1,134 @@
+package edu.texas.social_computing.hospitals;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.io.FileNotFoundException;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+public class RHRP {
+    private RHRP() {
+    }
+
+    /**
+     * Real HRP with Couples From http://www.nrmp.org/couples-match-videos/
+     * - Treat couples as a unit
+     * - Only make a match if the hospital is on resident's pref list and resident is on hospital's
+     * - If one partner has failed to match, increment the other's pref rank
+     * - If one partner has been kicked out due to being the worst, then break the other partner's match (it it has one)
+     * - Assumes that both partners have same pref list size
+     */
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents,
+                               Matching existingMatching) {
+        Matching m = existingMatching;
+
+        while (!freeResidents.isEmpty()) {
+            Resident currentResident = freeResidents.poll();
+            if (currentResident.getPreferences().isEmpty()) {
+                if (!currentResident.hasPartner()) {
+                    continue;
+                }
+                if (residentTable.getResidentById(currentResident.getPartnerId()).getPreferences().isEmpty()) {
+                    continue;
+                }
+                // resident and partner have to have the same pref list size
+                assert (false);
+            }
+
+            Hospital hospital = hospitalTable.getHospitalById(currentResident.getPreferences()
+                    .get(residentTable.getResidentRankProgress(currentResident)));
+            residentTable.incrementResidentRankProgress(currentResident);
+
+            tryToMatch(hospital,currentResident,residentTable,freeResidents,m);
+
+            // if the currentResident managed to get assigned then try to assign partner if he has one
+            if (m.hasAssignment(currentResident) && currentResident.hasPartner()) {
+                tryToMatch(hospital,residentTable.getResidentById(currentResident.getPartnerId()),
+                        residentTable,freeResidents,m);
+            }
+
+            // if currentResident has more applying to do, put back in Q
+            if (!m.hasAssignment(currentResident)) {
+                freeResidents.add(currentResident);
+                if (currentResident.hasPartner()) {
```

Comment:
id think we should only add partner to freeResidents if `!m.hasAssignment(partner)` is also true.

Reply:
I think that if either is not true, then both need to be unassigned and added back to the queue.

Reply:
so:
```
if(!m.hasAssignment(currentResident) || !m.hasAssignment(partner)) {
  m.unassign(currentResident);
  m.unassign(partner);
  freeResidents.add(currentResident);
  freeResidents.add(partner);
}

Reply:
This is much cleaner, but what if partner does not exist?

Reply:
Maybe too much work, but we _could_ add a dummy `NO_PARTNER` resident object with id "NO_PARTNER". The `NO_PARTNER` string in `Resident` gets changed from an empty string to "NO_PARTNER", and residents that do not have a partner get created with that string instead of empty string. `NO_PARTNER` would never get assigned anywhere, but could be entered into the Q, but would always be discarded right away. Or could conditionally add `partner` only if not equal to `NO_PARTNER`. That's just one idea :)